src/session.c

```diff
@@ -43,99 +178,160 @@ P11PROV_SESSION *p11prov_session_ref(P11PROV_SESSION *session)
-    if (ret != CKR_OK && ret != CKR_USER_ALREADY_LOGGED_IN) {
-        int retc;
-        P11PROV_raise(session->provctx, ret, "Error returned by C_Login");
-        retc = f->C_CloseSession(session->session);
-        if (retc != CKR_OK) {
-            P11PROV_raise(session->provctx, retc, "Failed to close session %lu",
-                          session->session);
+    if (login) {
+        /* Supports only USER login sessions for now */
+        ret = f->C_Login(session->session, CKU_USER, pin, pinlen);
+        if (ret != CKR_OK && ret != CKR_USER_ALREADY_LOGGED_IN) {
+            int retc;
+            P11PROV_raise(session->provctx, ret, "Error returned by C_Login");
+            retc = f->C_CloseSession(session->session);
+            if (retc != CKR_OK) {
+                P11PROV_raise(session->provctx, retc,
+                              "Failed to close session %lu", session->session);
+            }
+            session->session = CK_INVALID_HANDLE;
+            goto done;
         }
-        return ret;
     }
 
-    return CKR_OK;
+    ret = CKR_OK;
+
+done:
+    if (ret != CKR_OK) {
+        (void)__atomic_sub_fetch(&p->cur_sessions, 1, __ATOMIC_SEQ_CST);
+    }
+    return ret;
 }
 
 void p11prov_session_free(P11PROV_SESSION *session)
 {
+    int ref;
+
     if (session == NULL) {
         return;
     }
 
-    if (__atomic_sub_fetch(&session->refcnt, 1, __ATOMIC_SEQ_CST) != 0) {
+    ref = __atomic_sub_fetch(&session->refcnt, 1, __ATOMIC_SEQ_CST);
+    if (ref > 1) {
         return;
     }
+    if (ref < 1) {
+        /* only the pool should really free sessions.
+         * raise a warning */
+        P11PROV_raise(session->provctx, CKR_GENERAL_ERROR,
+                      "Potential double free in the calling code");
+        /* and restore the refcnt to 1 */
+        __atomic_store_n(&session->refcnt, 1, __ATOMIC_SEQ_CST);
+        return;
+    }
+    /* last ref means we go busy -> free */
+    if (ref == 1) {
+        P11PROV_SESSION_POOL *p = session->pool;
+        CK_ULONG cur_sessions;
+        bool ok;
+        int expected = 0;
+
+        if (p->login_session == session) {
+            __atomic_store_n(&session->pool->login_session, NULL,
+                             __ATOMIC_SEQ_CST);
+        }
 
-    if (session->session != CK_INVALID_HANDLE) {
-        CK_FUNCTION_LIST *f;
-        int ret;
-
-        f = p11prov_ctx_fns(session->provctx);
-        if (f != NULL) {
+        /* check if we used more than threshold sessions, in that case also
+         * close the session to avoid hogging all the sessions of a token */
+        cur_sessions = __atomic_load_n(&p->cur_sessions, __ATOMIC_SEQ_CST);
+        if (cur_sessions > p->limit_sessions) {
+            CK_FUNCTION_LIST *f;
+            int ret;
+
+            f = p11prov_ctx_fns(session->provctx);
+            if (f == NULL) {
+                /* FIXME: should we abort, crash and burn here? */
+                return;
+            }
```

Comment:
For me `p11prov_ctx_fns` should always succeed. `p11prov_ctx_fns` should assert / crash / panic if it can't return valid function list. This would clean up the code in various places.

Reply:
I thought about this a few time, but I *really* do not like when nested down in the bowel somewhere l;ibraries crash my application because they have a bug.
This can lave applications in bad states.
However what I could do is to mark the provider failed if there is ever an error like this and then refuse any further operation with it.

Reply:
See #53 for my take on this.